{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0a652e46-e86c-4ae4-9508-00bee9acd971",
   "metadata": {},
   "source": [
    "# L6c: Boltzamnn Gene Expression Logic into Flux Balance Analysis\n",
    "In this lecture, we'll introduce another model for the $u(...)$ and $w(...)$ functions that appear in the mRNA and protein equations. This approach is based upon building models of the logic driving gene expression for each individual promoter.\n",
    "The key concepts in the lecture are:\n",
    "* Fill me in."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbe1d85a-9402-4023-8d49-5673abfb62dd",
   "metadata": {},
   "source": [
    "## Review\n",
    "In the last lecture, introduced our first model of gene expression logic that relied on [boolean logic](https://en.wikipedia.org/wiki/Boolean_algebra). In this model, we formulate the mRNA and protein equations as:\n",
    "Let's investigate how we could describe gene regulation in flux balance analysis. Suppose the flux problem we were interested in was composed of enzymes encoded by the genes $\\mathcal{G}=1,2,\\dots, N$.\n",
    "The _action_ of each gene is described by two differential equations, one for mRNA concentration ($m_{j}$, units: `nmol/gDW`) and a second for the corresponding protein concentration ($p_{j}$, units: `nmol/gDW`):\n",
    "$$\n",
    "\\begin{align*}\n",
    "\t\\dot{m}_{j} &= r_{X,j}u_{j}\\left(\\dots\\right) - \\left(\\theta_{m,j}+\\mu\\right)\\cdot{m_{j}}+\\lambda_{j}\\quad{j=1,2,\\dots,N}\\\\\n",
    "\t\\dot{p}_{j} &= r_{L,j}w_{j}\\left(\\dots\\right) - \\left(\\theta_{p,j}+\\mu\\right)\\cdot{p_{j}}\n",
    "\\end{align*}\n",
    "$$\n",
    "Terms in the balances:\n",
    "* _Transcription_: The term $r_{X,j}u_{j}\\left(\\dots\\right)$ in the mRNA balance, which denotes the _regulated rate of transcription_ for gene $j$. This is \n",
    "the product of a _kinetic limit_ $r_{X,j}$ (units: `nmol/gDW-h`) and a transcription control function $0\\leq{u_{j}\\left(\\dots\\right)}\\leq{1}$ (dimensionless).\n",
    "The final term $\\lambda_{j}$ is the _unregulated expression rate_ of mRNA $j$ (units: `nmol/gDW-time`), i.e., this is the _leak_ expression rate.\n",
    "* _Translation_: The _regulated rate of translation_ of mRNA $j$, denoted by $r_{L,j}w_{j}$, is also the product of the\n",
    "kinetic limit of translation (units: `nmol/gDW-time`) and a translational control term $0\\leq{w_{j}\\left(\\dots\\right)}\\leq{1}$ (dimensionless).\n",
    "* _Degradation_: Lastly, $\\theta_{\\star,j}$ denotes the first-order rate constant (units: `1/time`) governing degradation of protein and mRNA, and $\\mu$ is the specific growth rate of the cell (units: `1/time`). We get the latter term using cell-specific concentration units (e.g., `nmol/gDW`).\n",
    "\n",
    "### Steady-State Concentrations\n",
    "At steady-state, the Let's show the steps to compute the steady-state mRNA concentration $m^{\\star}_{j}$:\n",
    "$$\n",
    "\\begin{align*}\n",
    "r_{X,j}u_{j}\\left(\\dots\\right) - \\left(\\theta_{m,j}+\\mu\\right)\\cdot{m_{j}}+\\lambda_{j} & = \\dot{m}_{j}\\\\\n",
    "r_{X,j}u_{j}\\left(\\dots\\right) - \\left(\\theta_{m,j}+\\mu\\right)\\cdot{m^{\\star}_{j}}+\\lambda_{j} &= 0 \\\\\n",
    "r_{X,j}u_{j}\\left(\\dots\\right) + \\lambda_{j} & = \\left(\\theta_{m,j}+\\mu\\right)\\cdot{m^{\\star}_{j}}\\\\\n",
    "\\frac{r_{X,j}u_{j}\\left(\\dots\\right) + \\lambda_{j}}{\\theta_{m,j}+\\mu} &= m^{\\star}_{j}\\quad\\text{for }j=1,2,\\dots,N\\quad\\blacksquare\n",
    "\\end{align*}\n",
    "$$\n",
    "Following the same steps, we can compute the steady-state protein concentration $p^{\\star}_{j}$:\n",
    "$$\n",
    "\\begin{equation*}\n",
    "p^{\\star}_{j} = \\frac{r_{L,j}w_{j}\\left(\\dots\\right)}{\\theta_{p,j}+\\mu}\\quad\\text{for }j=1,2,\\dots,N\\quad\\blacksquare\n",
    "\\end{equation*}\n",
    "$$\n",
    "\n",
    "If we had boolean descriptions for the control functions, we could estimate the steady-state mRNA and protein concentrations. Then the steady-state mRNA and protein concentration expressions in the `ON` case ($u_{j} = 1$ and $w_{j} = 1$) are given by:\n",
    "$$\n",
    "\\begin{align*}\n",
    "m^{\\star}_{j} &= \\frac{r_{X,j} + \\lambda_{j}}{\\theta_{m,j}+\\mu}\\qquad\\,p^{\\star}_{j} = \\frac{r_{L,j}}{\\theta_{p,j}+\\mu}\\\\\n",
    "\\end{align*}\n",
    "$$\n",
    "and in the `OFF` case with $\\lambda_{j}>0$:\n",
    "$$\n",
    "\\begin{align*}\n",
    "m^{\\star}_{j} &= \\frac{\\lambda_{j}}{\\theta_{m,j}+\\mu}\\qquad\\,p^{\\star}_{j} = \\frac{r_{L,j}}{\\theta_{p,j}+\\mu}\\\\\n",
    "\\end{align*}\n",
    "$$\n",
    "or in the `OFF` case if $\\lambda_{j}=0$: \n",
    "$$\n",
    "\\begin{align*}\n",
    "m^{\\star}_{j} &= 0\\qquad\\,p^{\\star}_{j} = 0\\\\\n",
    "\\end{align*}\n",
    "$$\n",
    "\n",
    "### Wrinkle: Going from protein to enzyme abundance\n",
    "Once we have $p^{\\star}_{j}$, we can compute the enzyme abundance $e_{j}$ in the system. However, there are three _base cases_ to consider:\n",
    "* __One to one__: If enzyme $e$ corresponds directly to protein $p$, then we can use the $p^{\\star}$ expression directly.\n",
    "* __Multisubunit__: If $e_{j}$ is a complex of different protein subunits, we can use the gene-protein-reaction (GPR) rules to compute the enzyme abundance $e_{j}$ as an `AND` combination. The `AND` rule requires all subunits to be expressed.\n",
    "* __Isoforms__: Alternatively, if $e_{j}$ is a single protein but there are multiple isoforms, we can use the GPR rules to compute the enzyme abundance $e_{j}$ as an `OR` combination.\n",
    "\n",
    "__Boolean logic in FBA example papers__:\n",
    "* [Covert MW, Palsson BO. Constraints-based models: regulation of gene expression reduces the steady-state solution space. J Theor Biol. 2003 Apr 7;221(3):309-25. doi: 10.1006/jtbi.2003.3071. PMID: 12642111.](https://pubmed.ncbi.nlm.nih.gov/12642111/)\n",
    "* [Orth JD, Fleming RM, Palsson BÃ˜. Reconstruction and Use of Microbial Metabolic Networks: the Core Escherichia coli Metabolic Model as an Educational Guide. EcoSal Plus. 2010 Sep;4(1). doi: 10.1128/ecosalplus.10.2.1. PMID: 26443778.](https://pubmed.ncbi.nlm.nih.gov/26443778/)\n",
    "* [Covert MW, Knight EM, Reed JL, Herrgard MJ, Palsson BO. Integrating high-throughput and computational data elucidates bacterial networks. Nature. 2004 May 6;429(6987):92-6. doi: 10.1038/nature02456. PMID: 15129285.](https://pubmed.ncbi.nlm.nih.gov/15129285/)\n",
    "\n",
    "Incorporating Boolean regulatory models, which are typically parameter-free, into flux balance analysis calculations (and ultimately metabolic design calculations) improves the ability of this type of mathematical model to simulate (predict) metabolic function.\n",
    "* __Could we do better?__ What if we could build a model of the logic driving gene expression for each individual promoter, that was a continuous function of the concentrations of the transcription factors, etc that bind to the promoter? "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c40859f5",
   "metadata": {},
   "source": [
    "### Discrete state promoter model\n",
    "Suppose a promoter $P$ can exist in $\\mathcal{S}$ possible _discrete_ microstates, and each microstate $s\\in\\mathcal{S}$ has some pseudo energy $\\epsilon_{s}$, where the _ground state_ $s_{1}\\in\\mathcal{S}$ has $\\epsilon_{1}=0$ (by definition). Some microstates will lead to expression (the ability to produce an mRNA molecule by transcription), while others will not. \n",
    "\n",
    "The probability that promoter $P$ is in microstate $s$ follows a [Boltzmann distribution](https://en.wikipedia.org/wiki/Boltzmann_distribution) which says:\n",
    "$$\n",
    "\\begin{align*}\n",
    "p_{s} & = \\frac{1}{Z} \\times f_{s}\\exp\\left(-\\beta\\epsilon_{s}\\right)\\qquad\\forall{s\\in\\mathcal{S}}\n",
    "\\end{align*}\n",
    "$$\n",
    "where $p_{s}$ is the probability of microstate $s\\in\\mathcal{S}$, \n",
    "$f_{s}$ is a system state specific factor where $f_{s}\\in\\left[0,1\\right]$,\n",
    "the $\\beta$ is the [thermodynamic beta](https://en.wikipedia.org/wiki/Thermodynamic_beta) and \n",
    "$Z$ is normalization factor (called the [Partiton function](https://en.wikipedia.org/wiki/Partition_function_(statistical_mechanics)) in the statistical physics community). We can find $Z$ using the summation law of discrete probolity $\\sum_{s}p_{s} = 1$ which gives:\n",
    "$$\n",
    "\\begin{align*}\n",
    "\\sum_{s\\in\\mathcal{S}}\\frac{1}{Z}\\times{f_{s}}\\exp\\left(-\\beta\\epsilon_{s}\\right) & = \\sum_{s\\in\\mathcal{S}}p_{s}\\\\\n",
    "\\frac{1}{Z}\\sum_{s\\in\\mathcal{S}}f_{s}\\exp\\left(-\\beta\\epsilon_{s}\\right) & = 1\\\\\n",
    "\\sum_{s\\in\\mathcal{S}}f_{s}\\exp\\left(-\\beta\\epsilon_{s}\\right) & = Z\\\\\n",
    "\\end{align*}\n",
    "$$\n",
    "which leads to the probability of microstate $s$:\n",
    "$$\n",
    "\\begin{align*}\n",
    "p_{s} & = \\frac{f_{s}\\exp\\left(-\\beta\\epsilon_{s}\\right)}{\\displaystyle \\sum_{s^{\\prime}\\in\\mathcal{S}}f_{i}\\exp\\left(-\\beta\\epsilon_{i}\\right)}\\qquad{s\\in\\mathcal{S}}\n",
    "\\end{align*}\n",
    "$$\n",
    "where $\\beta$ is the [thermodynamic beta](https://en.wikipedia.org/wiki/Thermodynamic_beta). Finally, we relate the probability that promoter $P$ is in microstate $s$ back to the $\\bar{u}\\left(\\dots\\right)$ control function by computing the overall probability that the desired event happens, e.g., promoter $P$ undergoes transcription. We know if $\\Omega = \\left\\{1,2,\\dots,\\mathcal{S}\\right\\}$, then we can define the subset $\\mathcal{A}\\subseteq\\Omega$ in which the desired event happens (in this case transcription). Given $\\mathcal{A}$, the $\\bar{u}\\left(\\dots\\right)$ function becomes:\n",
    "\n",
    "$$\\bar{u}=\\sum_{s\\in{\\mathcal{A}}}p_{s}$$\n",
    "\n",
    "where the control function $\\bar{u}$ is the sum of a regulated $u\\left(\\dots\\right)$ and unregulated $u^{\\dagger}\\left(\\dots\\right)$ component:\n",
    "\n",
    "$$\\bar{u} = u + u^{\\dagger}$$\n",
    "\n",
    "##### Example\n",
    "\n",
    "* [Moon TS, Lou C, Tamsir A, Stanton BC, Voigt CA. Genetic programs constructed from layered logic gates in single cells. Nature. 2012;491(7423):249-253. doi:10.1038/nature11516](https://www.ncbi.nlm.nih.gov/labs/pmc/articles/PMC3904217/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69fc7f9",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.11.2",
   "language": "julia",
   "name": "julia-1.11"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
